common/logger/src/security.rs

```diff
@@ -48,6 +48,9 @@ pub mod security_events {
     /// Consensus received an invalid proposal
     pub const INVALID_CONSENSUS_PROPOSAL: &str = "InvalidConsensusProposal";
 
+    /// Consensus received a proposal that is too far in the future
```

Comment:
maybe we should have two different logs here:

* proposal that's too far in the past 
* block (through the block API or in a proposal) that is too far in the future

Reply:
A proposal in the past would be rejected and that would be logged, already, no?

If the previous blocks are well formed, it should be bounded by its previous block:
https://github.com/libra/libra/blob/3bf241dd282c8c245da68efc259c11ff79e4a28c/consensus/consensus-types/src/block.rs#L240-L243
Or am I mixing things up?

Reply:
no you're right!

Reply:
Actually it is bounded by the previous block in the past, but it does not seem to enforce a minimum `delta` with the current time as it is said it should there:

https://github.com/libra/libra/blob/9bb8e3664485204e2f1224782a3700877ecde61e/consensus/consensus-types/src/block_data.rs#L57-L62

So if block A is agreed at time T and that the next block B is proposed at time T+ 1 with timestamp T+1,  it should be not be voted for if it's received after time T + delta, but I believe that currently it would be voted for.

Can I try and fix it by adding such a delta? I would do it in [verify_well_formed()](https://github.com/libra/libra/blob/d024369f7347570f7b9953c2cb5609af638e8c7d/consensus/consensus-types/src/proposal_msg.rs#L32) from the proposal_msg.rs, not in the block since it's really a condition on the proposal validation I'd say.
The problem is that currently the delta value is not formerly defined anywhere in the codebase, since this is not enforced.
What is a sensible value for delta in your opinion? 5 minutes ? 10 minutes ?

Reply:
I don't read this as the time difference between blocks! Rather, the time difference between a block being proposed and a vote. This is an information that you don't have here so you can't check for it. 

Also I'm not sure what we actually do. According to [the spec](https://github.com/libra/libra/blob/master/specifications/consensus/consensus.md#processing-a-proposalmsg) it seems like you can receive a proposal and vote on it even after you timed out (and broadcasted a timeout vote), as long as you haven't incremented your round (by receiving enough timeouts from other peers). I think I'm missing something here, as we would already have a timeout vote saved in cache and we would replace it with a real vote from safety rules...

Reply:
OK so we do bail early in `process_proposal` if it's passed the timeout, the spec should be updated with that : o https://github.com/libra/libra/blob/master/consensus/src/round_manager.rs#L481

Reply:
thanks for working on this, two clarifications:
1. the delta property is removed in the spec and the code comments need updated
2. >if you receive a proposal you'll vote on it, even if you've already timed out (and broadcasted a timeout vote) 

the spec aligns with the code, if a round is timed out, it must vote for a NIL block. so if a proposal comes late, it won't vote on the block because of the first voting rule (vote on increasing round).

Reply:
I don't think this is how I'm reading the spec, because `process_local_timeout` does not use safety rules in the spec so when safety rules is called via `process_proposal` it'll see this as the first time it votes.